evm/src/ampleforth/AmpleforthSwapAdapter.sol

```diff
@@ -0,0 +1,435 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import {ISwapAdapter} from "src/interfaces/ISwapAdapter.sol";
+import {IWAMPL} from "./IWAMPL.sol";
+import {ISTAMPL} from "./ISTAMPL.sol";
+import {IBillBroker} from "./IBillBroker.sol";
+
+/**
+ * @title AmpleforthSwapAdapter
+ *
+ * @notice An adapter contract that handles swaps between:
+ *         AMPL <> WAMPL, AMPL <> SPOT and SPOT <> USDC.
+ *
+ */
+contract AmpleforthSwapAdapter is ISwapAdapter {
+    using SafeERC20 for IERC20;
+
+    /// @notice Address of AMPL token.
+    address public constant AMPL = 0xD46bA6D942050d489DBd938a2C909A5d5039A161;
+    /// @notice Address of WAMPL token (wrapped AMPL).
+    address public constant WAMPL = 0xEDB171C18cE90B633DB442f2A6F72874093b49Ef;
+    /// @notice Address of SPOT token (AMPL senior perpetual).
+    address public constant SPOT = 0xC1f33e0cf7e40a67375007104B929E49a581bafE;
+    /// @notice Address of stAMPL contract (AMPL junior perpetual).
+    address public constant STAMPL = 0x82A91a0D599A45d8E9Af781D67f695d7C72869Bd;
+    /// @notice Address of USDC token.
+    address public constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    /// @notice Address of BillBroker contract.
+    address public constant BILL_BROKER =
+        0xA088Aef966CAD7fE0B38e28c2E07590127Ab4ccB;
+
+    /// @notice Constant that defines the limit factor for a POOL's reserve.
+    uint256 public constant RESERVE_LIMIT_FACTOR = 10;
+
+    /// @notice Constant that defines the limit factor for the STAMPL pool.
+    uint256 public constant STAMPL_AMPL_RESERVE_LIMIT_FACTOR = 50;
+
+    /// @inheritdoc ISwapAdapter
+    function swap(
+        bytes32 poolId,
+        address sellToken,
+        address buyToken,
+        OrderSide, /* side */
+        uint256 specifiedAmount
+    ) external override returns (Trade memory trade) {
+        // No-op if zero amount
+        if (specifiedAmount == 0) {
+            return trade;
+        }
+
+        uint256 gasBefore = gasleft();
+        address pool = address(bytes20(poolId));
+
+        if (pool == WAMPL) {
+            trade.calculatedAmount =
+                _swapWAMPL(sellToken, buyToken, specifiedAmount);
+        } else if (pool == STAMPL) {
+            trade.calculatedAmount =
+                _swapSTAMPL(sellToken, buyToken, specifiedAmount);
+        } else if (pool == BILL_BROKER) {
+            trade.calculatedAmount =
+                _swapBillBroker(sellToken, buyToken, specifiedAmount);
+        } else {
+            revert Unavailable("PoolNotRecognized");
+        }
+
+        trade.gasUsed = gasBefore - gasleft();
+        trade.price = Fraction(trade.calculatedAmount, specifiedAmount);
+
+        return trade;
```

Comment:
Since we have the named return variable in the prototype, we could omit this line.